aten/src/ATen/templates/Type.h

```diff
@@ -135,11 +135,23 @@ struct CAFFE2_API Type {
   /// Constructs the `TensorOptions` from a type and a `device_index`.
   TensorOptions options(int16_t device_index = -1) const {
     return TensorOptions().dtype(typeMeta())
-                          .device(backendToDeviceType(backend()), device_index)
+                          .device(device_type(), device_index)
                           .layout(layout())
                           .is_variable(is_variable());
   }
 
+  /// Constructs the `TensorOptions` from a type and a Device.  Asserts that
+  /// the device type matches the device type of the type.
+  TensorOptions options(optional<Device> device_opt) const {
+    if (!device_opt.has_value()) {
```

Comment:
Actually, it makes sense to me. A `Type` object semantically represents much of the same information that a `TensorOptions` has, except is doesn't hold a device index. So, logically, if you want to get a `TensorOptions` from a `Type`, you had better tell it the missing device index.

Actually, in this respect, `options(DeviceIndex)` makes more sense as a constructor, because it asks for the "minimum" information needed to do this completion. `options(Device)` needs the extra assert because it also gets information it doesn't care about.

But maybe you're saying, you don't see why we are manufacturing TensorOptions from Type in the first place? That is indeed a good question.